pkg/rasactl/rasactl.go

```diff
@@ -0,0 +1,323 @@
+	Log logr.Logger
+
+	// Spinner stores a spinner object.
+	Spinner *status.SpinnerMessage
+
+	// Namespace is a namespace name.
+	Namespace string
+
+	isRasaXRunning  bool
+	isRasaXDeployed bool
+
+	// CloudProvider stores a type of a detected cloud provider.
+	CloudProvider *cloud.Provider
+
+	// Flags stores the command flags.
+	Flags *types.RasaCtlFlags
+}
+
+// InitClients initializes clients.
+func (r *RasaCtl) InitClients() error {
+	r.Spinner = &status.SpinnerMessage{}
+	r.Spinner.New()
+
+	cloudProvider := &cloud.Provider{Log: r.Log}
+	cloudProvider.New()
+	r.CloudProvider = cloudProvider
+
+	kubernetesClient, err := k8s.New(
+		&k8s.Kubernetes{
+			Namespace:     r.Namespace,
+			Log:           r.Log,
+			CloudProvider: r.CloudProvider,
+			Flags:         r.Flags,
+		},
+	)
+	if err != nil {
+		return err
+	}
+	r.KubernetesClient = kubernetesClient
+
+	helmClient, err := helm.New(
+		&helm.Helm{
+			Log:           r.Log,
+			Namespace:     r.Namespace,
+			Spinner:       r.Spinner,
+			CloudProvider: r.CloudProvider,
+			Flags:         r.Flags,
+		},
+	)
+	if err != nil {
+		return err
+	}
+	helmClient.SetKubernetesBackendType(r.KubernetesClient.GetBackendType())
+	r.HelmClient = helmClient
+
+	dockerClient, err := docker.New(
+		&docker.Docker{
+			Namespace: r.Namespace,
+			Log:       r.Log,
+			Spinner:   r.Spinner,
+			Flags:     r.Flags,
+		},
+	)
+	if err != nil {
+		return err
+	}
+	r.DockerClient = dockerClient
+
+	err = r.GetKindControlPlaneNodeInfo()
+	return err
+}
+
+// SetNamespaceClients sets namespace for initialized clients.
+func (r *RasaCtl) SetNamespaceClients(namespace string) error {
+	r.Log.V(1).Info("Setting namespace for clients", "namespace", namespace)
+	r.KubernetesClient.SetNamespace(namespace)
+	r.DockerClient.SetNamespace(namespace)
+
+	err := r.HelmClient.SetNamespace(namespace)
+	return err
+}
+
+// CheckDeploymentStatus checks if a given deployment is deployed and running.
+//
+// It returns 'true' as the first value if the deployment is deployed, and 'true'
+// as the second value if the deployment is running.
+func (r *RasaCtl) CheckDeploymentStatus() (bool, bool, error) {
+	// Check if a Rasa X deployment is already installed and running
+	isRasaXDeployed, err := r.HelmClient.IsDeployed()
+	if err != nil {
+		return false, false, err
+	}
+	r.isRasaXDeployed = isRasaXDeployed
+
+	isRasaXRunning, err := r.KubernetesClient.IsRasaXRunning()
+	if err != nil {
+		return false, false, err
+	}
+
+	r.isRasaXRunning = isRasaXRunning
+
+	return isRasaXDeployed, isRasaXRunning, nil
+}
+
+func (r *RasaCtl) startOrInstall() error {
+	projectPath := r.Flags.Start.ProjectPath
+	// Install Rasa X
+	if !r.isRasaXDeployed && !r.isRasaXRunning {
+		if projectPath != "" || r.Flags.Start.Project {
+			if r.DockerClient.GetKind().ControlPlaneHost != "" {
+				if !r.Flags.Start.Project {
+					// check if the project path exists
+					if path, err := os.Stat(projectPath); err != nil {
+						if os.IsNotExist(err) {
```

Comment:
There's 6 nested `if`s at this point 😃  I recommend moving parts of this function to sub-functions in order to reduce the complexity a bit.

Reply:
Oh, yes, it's ugly, I was planning to go back to this 😄